lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+      ret = visitOperandAddptr(addptrOp, state, op.getLoc(), builder);
+    }
+    if (ret.failed()) {
+      op->emitRemark("Failed to rewrite yield op");
+      return failure();
+    }
+    initArgState.push_back(state);
+
+    // Verify that modulo state is not updated during the for loop
+    auto forState = knownPtrsFor[i];
+    for (auto i = 0; i < forState.getRank(); ++i) {
+      if (forState.modulos[i] != state.modulos[i]) {
+        // Special case, see comments in addState in dealing with modulos
+        if (i == 0 && forState.getRank() == 2) {
+          if (forState.modulos[1] == state.modulos[0] &&
+              forState.modulos[0] == state.modulos[1]) {
+            break;
+          }
+        }
+        assert(0);
+        op->emitRemark(
+            "PtrAnalysis: operand's modulo state changed within loop body");
+        return failure();
+      }
+    }
+  }
+
+  SmallVector<Value> operands;
+  for (auto opnd : op->getOperands()) {
+    auto mappedV = map.lookupOrNull(opnd);
+    if (mappedV) {
+      operands.push_back(mappedV);
+    } else {
+      operands.push_back(opnd);
+    }
+  }
+
+  // For each of the PtrSState recorded in the last step, extract value
+  // that correspond to offset and stride for each dimension and append
+  // them to yield operands.
+  for (auto state : initArgState) {
+    for (auto s : state.offsets) {
+      if (auto sIntAttr = getIntAttr(s)) {
+        auto constOp = builder.create<arith::ConstantOp>(
+            op.getLoc(), builder.getIndexAttr(sIntAttr.value()));
+        operands.push_back(constOp.getResult());
+      } else {
+        operands.push_back(s.get<Value>());
+      }
+    }
+
+    for (auto s : state.strides) {
+      assert(!getIntAttr(s) && "PtrSState strides for yield within for "
+                               "loop not expected to be attribute.");
+      operands.push_back(s.get<Value>());
+    }
+
+    if (state.getRank() == 0) {
+      operands.push_back(state.scalar);
+    }
+  }
+
+  auto newOp = builder.create<scf::YieldOp>(op->getLoc(), operands);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "new yield:";
+    newOp.getOperation()->print(llvm::dbgs(),
+                                OpPrintingFlags().printGenericOpForm());
+    llvm::dbgs() << "\n";
+  });
+
+  op->erase();
+  return success();
+}
+
+LogicalResult PtrAnalysis::rewriteLoadOp(triton::LoadOp op) {
+  auto ptr = map.lookupOrNull(op.getPtr());
+  auto mask = op.getMask();
+  auto other = op.getOther();
+  auto loc = op.getLoc();
+
+  if (!ptr) {
+    op->emitRemark("PtrAnalysis: pointer is not replace with tts.make_tptr so "
+                   "loadOp cannot be rewritten");
+    return failure();
+  }
+
+  if (!ptr.getType().isa<RankedTensorType>()) {
+    op->emitRemark("PtrAnalysis: scalar loadOp will not be rewritten");
+    return failure();
+  }
+
+  ArrayRef<OpFoldResult> dims;
+  MaskSState mstate;
+  Value scalarOther;
```

Comment:
nit: should move this inside the `if (mask)` block

Reply:
This is not obvious but since `dims` is an ArrayRef that can refer to `mstate`'s field, their lifetime need to be the same.